File: src/summarize/ProjectSummarizer.ts
import { Repository } from 'typeorm';
import { Project } from '../database/models/Project';
import { Summary } from '../database/models/Summary';
import { OpenAIClient } from './OpenAIClient';

export class ProjectSummarizer {
  private projectRepo: Repository<Project>;
  private summaryRepo: Repository<Summary>;
  private openaiClient: OpenAIClient;

  constructor(
    projectRepo: Repository<Project>,
    summaryRepo: Repository<Summary>,
    openaiApiKey?: string
  ) {
    this.projectRepo = projectRepo;
    this.summaryRepo = summaryRepo;
    this.openaiClient = new OpenAIClient(openaiApiKey);
  }

  async summarizeProject(projectId: number): Promise<Summary | null> {
    try {
      const project = await this.projectRepo.findOne({
        where: { id: projectId },
      });

      if (!project) {
        console.error(`Project ${projectId} not found`);
        return null;
      }

      // Check if summary already exists
      const existingSummary = await this.summaryRepo.findOne({
        where: { project_id: projectId },
      });

      if (existingSummary) {
        console.log(`Summary already exists for project ${projectId}`);
        return existingSummary;
      }

      // Generate summary and analysis
      const summaryText = await this.openaiClient.generateSummary(
        project.full_name,
        project.description || '',
        project.language
      );

      const analysis = await this.openaiClient.analyzeProject(
        project.full_name,
        project.description || '',
        project.language,
        project.stars
      );

      // Save summary
      const summary = this.summaryRepo.create({
        project_id: projectId,
        summary_text: summaryText,
        analysis: analysis,
      });

      await this.summaryRepo.save(summary);
      console.log(`Created summary for project ${projectId}`);

      return summary;
    } catch (error) {
      console.error(`Error summarizing project ${projectId}:`, error);
      throw error;
    }
  }

  async summarizeTrendingProjects(limit: number = 10): Promise<Summary[]> {
    try {
      // Get top trending projects without summaries
      const projects = await this.projectRepo
        .createQueryBuilder('project')
        .leftJoin('project.summaries', 'summary')
        .where('summary.id IS NULL')
        .orderBy('project.stars', 'DESC')
        .take(limit)
        .getMany();

      const summaries: Summary[] = [];

      for (const project of projects) {
        try {
          const summary = await this.summarizeProject(project.id);
          if (summary) {
            summaries.push(summary);
          }
          // Small delay to avoid rate limiting
          await new Promise((resolve) => setTimeout(resolve, 1000));
        } catch (error) {
          console.error(`Failed to summarize project ${project.id}:`, error);
        }
      }

      return summaries;
    } catch (error) {
      console.error('Error summarizing trending projects:', error);
      throw error;
    }
  }
}

File: src/summarize/project_summarizer.py
"""Generate AI summaries for GitHub projects"""
import logging
from typing import Dict, Any, Optional
from datetime import datetime
from sqlalchemy.orm import Session
from src.summarize.openai_client import OpenAIClient
from src.database.models import Project, Summary

logger = logging.getLogger(__name__)


class ProjectSummarizer:
    """Generate AI-powered summaries for GitHub projects"""

    SYSTEM_MESSAGE = """You are a technical analyst specializing in evaluating open-source GitHub projects.
Provide clear, concise, and informative summaries that help developers quickly understand a project's value and purpose."""

    SUMMARY_TEMPLATE = """Analyze this GitHub project and provide a comprehensive summary:

Project Name: {name}
Description: {description}
Programming Language: {language}
Stars: {stars}
URL: {url}

Please provide:
1. What this project does (2-3 sentences)
2. Usefulness rating (1-10 with brief justification)
3. Key technologies and frameworks used
4. Target audience and use cases
5. Notable features or advantages

Format your response as a structured analysis."""

    def __init__(self, db_session: Session, openai_client: OpenAIClient = None):
        """
        Initialize project summarizer

        Args:
            db_session: Database session
            openai_client: OpenAI client instance (creates new if None)
        """
        self.db = db_session
        self.openai_client = openai_client or OpenAIClient()

    def generate_summary(self, project_data: Dict[str, Any]) -> str:
        """
        Generate AI summary for a project

        Args:
            project_data: Dictionary containing project information

        Returns:
            Generated summary text
        """
        prompt = self.SUMMARY_TEMPLATE.format(
            name=project_data.get('name', 'Unknown'),
            description=project_data.get('description', 'No description'),
            language=project_data.get('language', 'Unknown'),
            stars=project_data.get('stars', 0),
            url=project_data.get('url', '')
        )

        logger.info(f"Generating summary for project: {project_data.get('name')}")

        try:
            summary = self.openai_client.generate_completion(
                prompt=prompt,
                system_message=self.SYSTEM_MESSAGE,
                max_tokens=600,
                temperature=0.7
            )
            return summary
        except Exception as e:
            logger.error(f"Error generating summary: {e}")
            raise

    def summarize_project(self, project_id: int) -> Summary:
        """
        Generate and save summary for a project by ID

        Args:
            project_id: Project database ID

        Returns:
            Created Summary object
        """
        # Get project from database
        project = self.db.query(Project).filter(Project.id == project_id).first()
        if not project:
            raise ValueError(f"Project with ID {project_id} not found")

        # Check if summary already exists
        existing_summary = self.db.query(Summary).filter(
            Summary.project_id == project_id
        ).first()

        if existing_summary:
            logger.info(f"Summary already exists for project {project.full_name}")
            return existing_summary

        # Prepare project data
        project_data = {
            'name': project.name,
            'description': project.description,
            'language': project.language,
            'stars': project.stars,
            'url': project.url
        }

        # Generate summary
        summary_text = self.generate_summary(project_data)

        # Save to database
        summary = Summary(
            project_id=project_id,
            summary_text=summary_text,
            analysis=summary_text,  # Can be parsed separately if needed
            created_at=datetime.now()
        )
        self.db.add(summary)
        self.db.commit()

        logger.info(f"Saved summary for project: {project.full_name}")
        return summary

    def batch_summarize(self, project_ids: list = None, limit: int = None) -> int:
        """
        Generate summaries for multiple projects

        Args:
            project_ids: List of project IDs to summarize. If None, summarize all projects
            limit: Maximum number of projects to summarize

        Returns:
            Number of summaries generated
        """
        # Get projects to summarize
        query = self.db.query(Project)

        if project_ids:
            query = query.filter(Project.id.in_(project_ids))

        # Exclude projects that already have summaries
        existing_summaries = self.db.query(Summary.project_id).all()
        existing_ids = [s[0] for s in existing_summaries]
        query = query.filter(~Project.id.in_(existing_ids))

        if limit:
            query = query.limit(limit)

        projects = query.all()

        logger.info(f"Generating summaries for {len(projects)} projects")

        count = 0
        for project in projects:
            try:
                self.summarize_project(project.id)
                count += 1
            except Exception as e:
                logger.error(f"Failed to summarize project {project.full_name}: {e}")
                continue

        logger.info(f"Successfully generated {count} summaries")
        return count
